docs/proposals/server-side-pagination.md

```diff
@@ -0,0 +1,180 @@
+
+This is where to call out areas of the design that require closure before deciding to implement the
+design.
+
+
+## Summary
+
+The Argo CD API server currently returns all applications in a single response. This can be a performance
+bottleneck when there are a large number of applications. This proposal is to introduce server side pagination
+for the Applications List and Watch APIs.
+
+## Motivation
+
+The main motivation for this proposal it to improve the Argo CD UI responsiveness when there are a large number
+of applications. The API server memory usage increases with the number of applications however this is not critical
+and can be mitigated by increasing memory limits for the API server deployment. The UI however becames unresponsive
+even on a powerful machine when the number of applications increases 2000. The server side pagination will allow
+to reduce amount of data returned by the API server and improve the UI responsiveness.
+
+### Goals
+
+* Support server side pagination for Applications List and Watch APIs
+
+* Leverage pagination in the Argo CD UI to improve responsiveness
+
+* Leverage pagination in the Argo CD CLI to improve performance and reduce load on the API server
+
+### Non-Goals
+
+* The API Server is known to use a lot of CPU while applying very large RBAC policies to a large number of applications.
+  Even with pagination API still need to apply RBAC policies to return "last page" response. So the issueis not addressed by this proposal.
+
+## Proposal
+
+**Pagination Cursor**
+
+It is proposed to add `offset` and `limit` fields for pagination support in Application List API.
+The The Watch API is a bit more complex. Both Argo CD user interface and CLI are relying on the Watch API to display real time updates of Argo CD applications.
+The Watch API currently supports filtering by a project and an application name. In order to effectively
+implement server side pagination for the Watch API we cannot rely on the order of the applications returned by the API server. Instead of
+relying on the order it is proposed to rely on the application name and use it as a cursor for pagination. Both the Applications List and Watch
+APIs will be extended with the optional `minName` and `maxName` fields. The `minName` field will be used to specify the application name to start from
+and the `maxName` field will be used to specify the application name to end at. The fields should be added to the `ApplicationQuery` message
+which is used as a request payload for the Applications List and Watch APIs.
+
+```proto
+message ApplicationQuery { 
+  // ... existing fields
+  // New proto fields for server side pagination
+	// the application name to start from (app with min name is included in response)
+	optional string minName = 9;
+	// the application name to end at (app with max name is included in response)
+	optional string maxName = 10;
+  	// offset
+	optional int64 offset = 18;
+	// limit
+	optional int64 limit = 19;
```

Comment:
Late on a Friday, I'm having trouble understanding the usage of this API. It seems to me that it makes sense to use _either_ offset/limit _or_ minName/maxName. Are they meant to be used in conjunction?

Reply:
That makes sense! But +1, maybe just a bit more clarification.

Reply:
Being familiar with https://docs.github.com/en/rest/using-the-rest-api/using-pagination-in-the-rest-api?apiVersion=2022-11-28 mechanism, I am not sure to understand why minName/maxName are required. 

Are these 2 used for server side filtering? If so, maybe move them to the next section to avoid confusion.

Reply:
@agaudreault Do we need to make that change to merge or can we close this conversation?

Reply:
Not a  blocker, more of an open question that might come back during the implementation. We'll see when we get there 👍